packages/@lwc/engine/src/framework/hot-swaps.ts

```diff
@@ -0,0 +1,238 @@
+        // they will get re-associated once these instances are rehydrated.
+        list.clear();
+    }
+    return canRefreshAllInstances;
+}
+
+export function getTemplateOrSwappedTemplate(tpl: Template): Template {
+    if (process.env.NODE_ENV === 'production') {
+        // this method should never leak to prod
+        throw new ReferenceError();
+    }
+    const visited: Set<Template> = new Set();
+    while (swappedTemplateMap.has(tpl) && !visited.has(tpl)) {
+        visited.add(tpl);
+        tpl = swappedTemplateMap.get(tpl)!;
+    }
+    return tpl;
+}
+
+export function getComponentOrSwappedComponent(Ctor: ComponentConstructor): ComponentConstructor {
+    if (process.env.NODE_ENV === 'production') {
+        // this method should never leak to prod
+        throw new ReferenceError();
+    }
+    const visited: Set<ComponentConstructor> = new Set();
+    while (swappedComponentMap.has(Ctor) && !visited.has(Ctor)) {
+        visited.add(Ctor);
+        Ctor = swappedComponentMap.get(Ctor)!;
+    }
+    return Ctor;
+}
+
+export function getStyleOrSwappedStyle(style: StylesheetFactory): StylesheetFactory {
+    if (process.env.NODE_ENV === 'production') {
+        // this method should never leak to prod
+        throw new ReferenceError();
+    }
+    const visited: Set<StylesheetFactory> = new Set();
+    while (swappedStyleMap.has(style) && !visited.has(style)) {
+        visited.add(style);
+        style = swappedStyleMap.get(style)!;
+    }
+    return style;
+}
+
+export function setActiveVM(vm: VM) {
+    if (process.env.NODE_ENV === 'production') {
+        // this method should never leak to prod
+        throw new ReferenceError();
+    }
+    // tracking active component
+    const Ctor = vm.def.ctor;
+    let componentVMs = activeComponents.get(Ctor);
+    if (isUndefined(componentVMs)) {
+        componentVMs = new Set();
+        activeComponents.set(Ctor, componentVMs);
+    }
+    // this will allow us to keep track of the hot components
+    componentVMs.add(vm);
+
+    // tracking active template
+    const tpl = vm.cmpTemplate;
+    let templateVMs = activeTemplates.get(tpl);
+    if (isUndefined(templateVMs)) {
+        templateVMs = new Set();
+        activeTemplates.set(tpl, templateVMs);
+    }
+    // this will allow us to keep track of the templates that are
+    // being used by a hot component
+    templateVMs.add(vm);
+
+    // tracking active styles associated to template
+    const stylesheets = tpl.stylesheets;
+    if (!isUndefined(stylesheets)) {
+        stylesheets.forEach((stylesheet) => {
+            // this is necessary because we don't hold the list of styles
+            // in the vm, we only hold the selected (already swapped template)
+            // but the styles attached to the template might not be the actual
+            // active ones, but the swapped versions of those.
+            stylesheet = getStyleOrSwappedStyle(stylesheet);
```

Comment:
@ravijayaramappa this is not very elegant. I think we should start storing the styles into the vm, after all the swapping and computations. then here we don't need to do this.